flink-ml-lib/src/main/java/org/apache/flink/ml/feature/bucketizer/Bucketizer.java

```diff
@@ -0,0 +1,179 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.feature.bucketizer;
+
+import org.apache.flink.api.common.functions.FlatMapFunction;
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.ml.api.Transformer;
+import org.apache.flink.ml.common.datastream.TableUtils;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Bucketizer is a transformer that maps multiple columns of continuous features to multiple columns
+ * of discrete features, i.e., buckets IDs.
+ */
+public class Bucketizer implements Transformer<Bucketizer>, BucketizerParams<Bucketizer> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public Bucketizer() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        String[] inputCols = getInputCols();
+        String[] outputCols = getOutputCols();
+        Double[][] splitsArray = getSplitsArray();
+
+        Preconditions.checkArgument(inputCols.length == outputCols.length);
+        Preconditions.checkArgument(inputCols.length == splitsArray.length);
+        for (Double[] splits : splitsArray) {
+            Preconditions.checkArgument(
+                    splits.length >= 3,
+                    "Illegal value for "
+                            + BucketizerParams.SPLITS_ARRAY
+                            + ". See param "
+                            + BucketizerParams.SPLITS_ARRAY
+                            + " for details.");
+            for (int j = 1; j < splits.length; j++) {
+                Preconditions.checkArgument(
+                        splits[j] > splits[j - 1],
+                        "Illegal value for "
+                                + BucketizerParams.SPLITS_ARRAY
+                                + ". See param "
+                                + BucketizerParams.SPLITS_ARRAY
+                                + " for details.");
+            }
+        }
```

Comment:
Shall we move this part of logic to a `ParamValidator` class? This class can be placed inside `BucketizerParams`.

Moreover, We would always need to do some checks in ParamValidator, and to do the other checks in `fit()`/`transform()`. I think we should place a validation in ParamValidator if it is generally applicable to all subclasses that inherit this parameter. If the validation needs to be computed across parameters, or depends on the input Tables, then it should be placed in `fit()`/`transform()`.

Reply:
Good point. I have move the checks to `BucketizerParams`.